### PR TITLE
Add action to format code

### PR DESCRIPTION
**Pull request by ErikSchierboom**
Depends on https://github.com/exercism/problem-specifications/pull/1917 to be merged

This is almost a carbon copy of the [workflow that the JavaScript track uses](https://github.com/exercism/javascript/blob/main/.github/workflows/action-format.yml).

@iHiD This repo now needs access to the SSH deploy key, like the `javascript` track.

### CODE REVIEW

**Comment by SaschaMann** [comment]
> @iHiD This repo now needs access to the GITHUB_TOKEN secret, like the javascript track.

It has that by default, doesn't it? The secret it needs in addition to the defaults is an SSH deploy key so that pushing triggers CI.

**Comment by ErikSchierboom** [comment]
Ah, makes sense.

File: .github/workflows/action-format.yml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+          else
+            echo "::debug::Setting up repo using HTTPS"
+            HEAD_REPO=$(jq -r '.head.repo.clone_url | sub("https://"; "https://x-access-token:${{ secrets.GITHUB_TOKEN }}@")' "$PR_DATA")
+          fi
+
+          git clone $HEAD_REPO .
+          git checkout -b "$HEAD_REF" "origin/$HEAD_REF"
+
+      - name: Use Node.js LTS (16.x)
+        uses: actions/setup-node@1f8c6b94b26d0feae1e387ca63ccbdc44d27b561
```
Comment (wolf99) [inline .github/workflows/action-format.yml:64]: Any updates needed for `.github/dependabot.yml` for the `uses` actions (such as this one, actions/cache, actions/github-script) ?

File: .github/workflows/action-format.yml [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          script: |
+            github.issues.createComment({
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: 'The "Format code" action has finished running.'
+            })
+
+      - name: 'Post reminder to trigger build manually'
```
Comment (wolf99) [inline .github/workflows/action-format.yml:118]: A question for my own understanding: in what use case does this step run?

It seems like when it is a fork, this is catered for (e.g. ln # 55 - 58), and there is nothing stopping the other steps running when it is a fork... yet this step seems to imply that the job should not work for a fork?

Comment (SleeplessByte) [reply .github/workflows/action-format.yml:118]: It actually says it in the message!

> For security reasons, `/format` does not trigger CI builds when the PR has been submitted from a fork. If checks were not passing due to code format, trigger a build to make the required checks pass, through one of the following ways:
> 
> - Push an empty commit to this branch: `git commit --allow-empty -m "Trigger builds"`.
> - Close and reopen the PR.
> - Push a regular commit to this branch.

Basically when you _write_ commits (so you use the SECRET token, which we do in the format _write_ action) the CI doesn't trigger. So it's not that you cannot run this action on a forked-pr (you can!), it's that _after_ running this action, it won't re-trigger a CI run from the commit that this action writes.

File: .github/workflows/action-format.yml [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+
+      - name: Install dependencies
+        run: yarn install --pure-lockfile
+
+      - name: Verify that canonical-data.json files adhere to their json schema
+        run: yarn format-json
```
Comment (SleeplessByte) [inline .github/workflows/action-format.yml:86]: This doesn't verify. This writes. Right?

File: .github/workflows/action-format.yml (hunk 2 above)
Comment (SaschaMann) [reply .github/workflows/action-format.yml:118]: > Basically when you write commits (so you use the SECRET token, which we do in the format write action) the CI doesn't trigger. So it's not that you cannot run this action on a forked-pr (you can!), it's that after running this action, it won't re-trigger a CI run from the commit that this action writes.

This is only true for commits written using the default GitHub Actions token. When the PR is from a branch within this repo, we can use an SSH deploy key to push to it and avoid this issue. That's why the workflow has two cases that are handled differently (fork vs non-fork).

Comment (SleeplessByte) [reply .github/workflows/action-format.yml:118]: Oh yeah, there's that :)

Comment (wolf99) [reply .github/workflows/action-format.yml:118]: Great, thanks for clarifying folks!

File: .github/workflows/action-format.yml (hunk 1 above)
Comment (ErikSchierboom) [reply .github/workflows/action-format.yml:64]: No, this should be covered by https://github.com/exercism/problem-specifications/blob/main/.github/dependabot.yml#L6. 
See https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-the-dependency-graph#supported-package-ecosystems

File: .github/workflows/action-format.yml (hunk 3 above)
Comment (ErikSchierboom) [reply .github/workflows/action-format.yml]: Correct. I've updated.

**Review by wolf99 — APPROVED** [review]

**Review by kotp — APPROVED** [review]

**Review by SaschaMann — APPROVED** [review]

**Comment by iHiD** [comment]
I think I've done it but I don't really have any idea what I'm doing :)

**Comment by SaschaMann** [comment]
@iHiD in that case, could you approve this PR so we can merge it and we'll try it out in a test PR? :)

**Review by iHiD — APPROVED** [review]

**Pull request merged**